### PR TITLE
fixup! Add X11DisplayManager to get display information

### PR DESCRIPTION
**Pull request by jkim-julie**
This patch removed OwnedDisplaySnapshot because it doesn't need to
own DisplayMode. DisplaySnapshot is created with
DisplaySnapshot::DisplayModeList, which is the list has DisplayModes.
Whenever DisplayMode is added to the list, it's cloned and kept in
the list and every DisplayMode should be in the list.

### CODE REVIEW

**Comment by jkim-julie** [comment]
@msisov, there is a crash with 'views_mus_unittests' on ChromeOS with X11DisplayManagerOzone.

This patch sets |native_mode| for DisplaySnapshot, because when there is no |selected_mode| for DisplayId, DisplayConfigurator takes |native_mode|.
If there is no |selected_mode| or |native_mode|, GetDisplayLayout() is failed, DisplayConfigurator::OnConfigured() has no |success|, observer.OnDisplayModeChangeFailed() is called and it goes to 'ScreenManagerOzoneInternal::PostDisplayConfigurationChange()'. But |touch_transform_controller_| is not created yet on the sequence and we meet a crash.

Briefly, the sequence is below.
Service::StartDisplayInit()
->ScreenManagerOzoneInternal::Init() [1]
->DisplayConfigurator::ForceInitialConfigure() (https://cs.chromium.org/chromium/src/services/ui/display/screen_manager_ozone_internal.cc?sq=package:chromium&q=screen_manager_ozone_&g=0&l=158)
->UpdateDisplayConfigurationTask::Run()
->X11NativeDisplayDelegate::GetDisplays()
->X11DisplayManagerOzone::GetDisplaysSnapshot()
->UpdateDisplayConfigurationTask::OnDisplaysUpdated()
->UpdateDisplayConfigurationTask::EnterState() (It calls 'DisplayConfigurator::DisplayLayoutManagerImpl::GetDisplayLayout()' but if |state->selected_mode| is empty, it returns false. https://cs.chromium.org/chromium/src/ui/display/manager/display_configurator.cc?sq=package:chromium&q=display_confi&g=0&l=279)
->UpdateDisplayConfigurationTask::OnStateEntered()
->FinishConfiguration()
->DisplayConfigurator::OnConfigured()
->NotifyDisplayStateObservers()
->DisplayChangeObserver::OnDisplayModeChangeFailed()
->DisplayChangeObserver::OnDisplayModeChanged()
->DisplayManager::OnNativeDisplaysChanged()
->DisplayManager::UpdateDisplaysWith()
->ScreenManagerOzoneInternal::PostDisplayConfigurationChange() (Crashed because it doesn't create |touch_transform_controller_| yet at [1].)

With this flow, It has 3 problems now.
1) DisplayMode should be in the list in DisplaySnapshot.
2) DisplaySnapshot should have native_mode at least. FakeDisplaySnapshot also has it. (https://cs.chromium.org/chromium/src/ui/display/manager/fake_display_snapshot.cc?sq=package:chromium&q=fake_display_sna&g=0&l=156)
3) Even if DisplayConfigurator gets fail, it sholdn't be crashed.

This patch fixes the problems above and it implements 'X11NativeDisplayDelegate::Configure', which has same behaviour as FakeDisplayDelegate::Configure().

PTAL.

File: services/ui/display/screen_manager_ozone_internal.cc [hunk 1]
```diff
@@ -320,7 +320,8 @@ void ScreenManagerOzoneInternal::PostDisplayConfigurationChange() {
     }
   }
 
-  touch_transform_controller_->UpdateTouchTransforms();
+  if (touch_transform_controller_)
```
Comment (msisov) [inline services/ui/display/screen_manager_ozone_internal.cc:323]: why do you need to change this? This is mus/mash related.

File: ui/ozone/platform/x11/x11_display_manager_ozone.cc [hunk 2]
```diff
@@ -23,34 +24,36 @@ namespace {
...
+    gfx::Rect bounds,
+    gfx::ColorSpace color_space) {
+  display::DisplaySnapshot::DisplayModeList modes;
+  std::unique_ptr<display::DisplayMode> display_mode =
       std::make_unique<display::DisplayMode>(
           gfx::Size(bounds.width(), bounds.height()), false, default_refresh);
-  snapshot->set_current_mode(current_mode.get());
-  return std::make_pair(std::move(snapshot), std::move(current_mode));
+  modes.push_back(std::move(display_mode));
+  const display::DisplayMode* mode = modes.back().get();
```
Comment (msisov) [inline ui/ozone/platform/x11/x11_display_manager_ozone.cc:36]: |display_mode| is owned by |modes|. Where |modes| are only defined within the scope of this method. This shall result in undefined behavior later. Why can't you just store the current |modes| and the snapshots as WaylandOutput does?

File: ui/ozone/platform/x11/x11_native_display_delegate.cc [hunk 3]
```diff
@@ -42,7 +49,31 @@ void X11NativeDisplayDelegate::Configure(const display::DisplaySnapshot& output,
...
+    }
+  }
+
+  configure_callbacks_.push(
+      base::BindOnce(std::move(callback), configure_success));
+
+  // Start the timer if it's not already running. If there are multiple queued
+  // configuration requests then ConfigureDone() will handle starting the
+  // next request.
+  if (!configure_timer_.IsRunning()) {
```
Comment (msisov) [inline ui/ozone/platform/x11/x11_native_display_delegate.cc]: I can't understand why the timer needed. Can you explain more?

File: services/ui/display/screen_manager_ozone_internal.cc (hunk 1 above)
Comment (jkim-julie) [reply services/ui/display/screen_manager_ozone_internal.cc:323]: If you look at the sequence described at the previous comment, the crash happened with 'views_mus_unittests', it's started at https://cs.chromium.org/chromium/src/services/ui/display/screen_manager_ozone_internal.cc?q=ScreenManagerOzoneInternal::Init()&sq=package:chromium&g=0&l=158 and |touch_transform_controller_| wasn't created yet. So, accessing without checking caused crash.

File: ui/ozone/platform/x11/x11_display_manager_ozone.cc (hunk 2 above)
Comment (jkim-julie) [reply ui/ozone/platform/x11/x11_display_manager_ozone.cc:36]: When it creates display::DisplaySnapshot,  it deliver |modes| as a parameter and DisplaySnapshot keeps it. I refereed to others, how to create display::DisplaySnapshot, https://cs.chromium.org/chromium/src/ui/display/mojo/display_snapshot_struct_traits.cc?sq=package:chromium&g=0&l=131 or https://cs.chromium.org/chromium/src/ui/ozone/platform/drm/common/drm_util.cc?sq=package:chromium&q=drm_uti&g=0&l=472.

And whenever DisplaySnapshot::add_mode() is called, it clones every display mode and keeps it.

FYI, I think we should have this mechanism for headless as well and I was supposed to upstream it after confirm this change but it's already reverted today with the same reason. https://chromium-review.googlesource.com/c/chromium/src/+/1094851/5/ui/ozone/platform/headless/headless_native_display_delegate.cc#35

File: ui/ozone/platform/x11/x11_native_display_delegate.cc (hunk 3 above)
Comment (jkim-julie) [reply ui/ozone/platform/x11/x11_native_display_delegate.cc]: I refereed to FakeDisplayDelegate::Configure and it adds delay to call callback.

File: ui/ozone/platform/x11/x11_display_manager_ozone.cc (hunk 2 above)
Comment (msisov) [reply ui/ozone/platform/x11/x11_display_manager_ozone.cc:36]: ah, I haven't noticed it has also taken modes.

File: services/ui/display/screen_manager_ozone_internal.cc (hunk 1 above)
Comment (msisov) [reply services/ui/display/screen_manager_ozone_internal.cc:323]: ok

File: ui/ozone/platform/x11/x11_native_display_delegate.cc (hunk 3 above)
Comment (msisov) [reply ui/ozone/platform/x11/x11_native_display_delegate.cc]: Here is the patch, which adds configuration delay to a fake display - https://codereview.chromium.org/2527263002 .

I still don't understand why you think X11 display needs it? It's not fake. You must use X APIs for configuration then and avoid having this kind of synthetic delays.

Comment (msisov) [reply ui/ozone/platform/x11/x11_native_display_delegate.cc]: Or leave it NOTIMPLEMENTED

**Comment by jkim-julie** [comment]
Hi, thanks for the review. I removed timer from X11NativeDisplayDelegate::Configure() and added NOTIMPLEMENTED(). PTAL.

File: ui/ozone/platform/x11/x11_native_display_delegate.cc [hunk 4]
```diff
@@ -94,4 +99,5 @@ void X11NativeDisplayDelegate::OnOutputReadyForUse() {
   for (display::NativeDisplayObserver& observer : observers_)
     observer.OnConfigurationChanged();
 }
+
```
Comment (msisov) [inline ui/ozone/platform/x11/x11_native_display_delegate.cc]: nit: remove this line

File: ui/ozone/platform/x11/x11_display_manager_ozone.h [hunk 5]
```diff
@@ -7,20 +7,19 @@
...
 #include <stdint.h>
 
-#include <map>
-#include <memory>
-
-#include "base/cancelable_callback.h"
-#include "base/macros.h"
-#include "ui/display/screen.h"
-#include "ui/display/types/display_snapshot.h"
+#include "base/memory/weak_ptr.h"
```
Comment (msisov) [inline ui/ozone/platform/x11/x11_display_manager_ozone.h]: is it needed?

File: ui/ozone/platform/x11/x11_display_manager_ozone.cc [hunk 6]
```diff
@@ -23,34 +24,36 @@ namespace {
...
+      display_id, gfx::Point(bounds.x(), bounds.y()),
+      gfx::Size(bounds.width(), bounds.height()),
+      display::DisplayConnectionType::DISPLAY_CONNECTION_TYPE_NONE, false,
+      false, false, color_space, "", base::FilePath(), std::move(modes),
+      std::vector<uint8_t>(), mode, mode, 0, 0, gfx::Size());
 }
 
 void BuildFallbackDisplayList(
-    X11DisplayManagerOzone::OwnedDisplaySnapshot& output) {
+    std::vector<std::unique_ptr<display::DisplaySnapshot>>& output) {
```
Comment (msisov) [inline ui/ozone/platform/x11/x11_display_manager_ozone.cc]: AFAIK, current agreement is to pass parameters by pointers if they are going to be changed. Please, check chromium-style guide. Or you can change the return type and build this vector here and return it.

**Review by msisov — APPROVED** [review]
lgtm with nits

**Comment by jkim-julie** [comment]
Thanks for the comment. I updated the patch.

**Comment by msisov** [comment]
lgtm

**Pull request merged**